hal/lib/athena/CallbackVector.h

```diff
@@ -0,0 +1,116 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) FIRST 2016. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <vector>
+
+namespace hal {
+// Vector which provides an integrated freelist for removal and reuse of
+// individual elements.
+template <typename T>
+class CallbackVector {
+  struct private_init {};
+
+ public:
+  typedef typename std::vector<T>::size_type size_type;
+
+  // Constructor for creating copies of the vector
+  CallbackVector(const CallbackVector* copyFrom, const private_init&);
+
+  // Delete all default constructors so they cannot be used
+  CallbackVector& operator=(const CallbackVector&) = delete;
+  CallbackVector() = delete;
+  CallbackVector(const CallbackVector&) = delete;
+
+  // Create a new vector with a single callback inside of it
+  CallbackVector(T callback, unsigned int* newUid);
+
+  size_type size() const { return m_vector.size(); }
+  T& operator[](size_type i) { return m_vector[i]; }
```

Comment:
Is non-const `operator[]` ever used?  From what I can tell this class should essentially be immutable.